src/main/java/org/folio/uk/service/UserService.java

```diff
@@ -98,21 +108,50 @@ public Optional<User> getUser(UUID id) {
     return usersClient.lookupUserById(id);
   }
 
-  public CompositeUser getUserBySelfReference(List<IncludedField> include, boolean expandPermissions) {
+  public CompositeUser getUserBySelfReference(List<IncludedField> include, boolean expandPermissions,
+                                              boolean overrideUser) {
     log.info("Retrieving user by self reference with parameters: include = {}, expandPermissions = {}",
       () -> StringUtils.join(emptyIfNull(include), ", "), () -> expandPermissions);
 
     var userId = getUserId();
-
     var user = usersClient.lookupUserById(userId)
       .orElseThrow(() -> new EntityNotFoundException("User was Not Found with: id = " + userId));
 
-    return new CompositeUser()
-      .user(user)
+    // TODO Switch overrideUser back to assert true
+    if (Boolean.TRUE.equals(keycloakFederatedAuthProperties.isEnabled()) && Boolean.FALSE.equals(overrideUser)) {
+      var originalTenantIdOptional = user.getCustomFields().entrySet().stream()
+        .filter(entry -> entry.getKey().equalsIgnoreCase(ORIGINAL_TENANT_ID_CUSTOM_FIELD))
+        .map(entry -> (String) entry.getValue()).findFirst();
+      if (originalTenantIdOptional.isPresent()) {
+        return getRealUserByReference(user, originalTenantIdOptional.get(), expandPermissions);
+      }
+    }
+
+    return new CompositeUser().user(user)
       .permissions(fetchPermissionUser(userId, expandPermissions))
       .servicePointsUser(fetchServicePointUser(userId));
   }
 
+  private CompositeUser getRealUserByReference(User shadowUser, String originalTenantId, boolean expandPermissions) {
+    log.info("Overriding self reference to use a real shadowUser with: tenant = {}", originalTenantId);
+    if (StringUtils.isEmpty(shadowUser.getUsername())) {
+      throw new IllegalStateException("Username was Is Empty: id = " + shadowUser.getId());
+    }
+
+    try (var ignored = new FolioExecutionContextSetter(
+      TenantContextUtils.prepareContextForTenant(originalTenantId, folioModuleMetadata, folioExecutionContext))) {
+      var username = shadowUser.getUsername().substring(0, shadowUser.getUsername().length() - RANDOM_STRING_COUNT);
+      var realUser = usersClient.query("username==" + username, 1)
```

Comment:
can be simplified, we can found user by userId instead of username. All users(real and shadow copies) have the same userId in all tenants